data-structure/jqueue.php

```diff
@@ -0,0 +1,71 @@
+<?php
+class jQueue extends jDataStructures
+{
+    public function __construct($data)
+    {
+        parent::__construct($data);
+    }
+
+    /**
+     * put the element to the first position
+     * @param mixed $elem
+     * @return jQueue
+     */
+    public function push($elem)
+    {
+        array_unshift($this->_data, $elem);
+        return $this;
+    }
+
+    /**
+     * returns the element in the first position
+     * @throws Exception
+     * @return mixed
+     */
+    public function peek()
+    {
+        if ($this->isEmpty()) {
+            throw new Exception('The Queue is Empty');
+        }
+
+        return $this->_data[array_key_first($this->_data)];
+    }
+
+    /**
+     * remove the first element and returns it
+     * @throws Exception
+     * @return mixed
+     */
+    public function pop()
+    {
+        if ($this->isEmpty()) {
+            throw new Exception('The Queue is Empty');
+        }
+
+        return array_shift($this->_data);
```

Comment:
`array_shift` has o(n) time complexity. Rather than using this can we use a pointer that will point to the current index? it will start from 0. Every time you call `pop()` :
 -  it return the data with current index
 -  `unset()` the current index
 - increment the pointer value